### PR TITLE
[aes] Use sparse state encodings for FSMs, trigger alert for invalid states

### PR DESCRIPTION
**Pull request by vogelpi**
This PR changes all FSMs inside AES to use sparse state encodings with a minimum Hamming distance of 3 between states. If any FSM enters an invalid state, the AES unit locks and an alert is triggered. The cipher core is prevented from finishing the current operation or starting a new operation. Write to the control register are ignored. The AES unit needs to be reset.

This PR also renames the alert signals. All minor alerts (currently just update errors in shadow registers) are collected in `minor`. Fatal alerts such as storage errors in shadow registers, entering invalid FSM states trigger a `major` alert.

This PR only targets the FSMs. Further PRs will follow to also protect e.g. the round counter and mux signals.

### CODE REVIEW

**Comment by vogelpi** [comment]
FYI @cindychip , @rasmus-madsen , @weicaiyang, I've renamed the alert signals to collect multiple alert sources in a minor and major alert.

File: hw/ip/csrng/rtl/csrng_block_encrypt.sv
Comment (vogelpi) [inline hw/ip/csrng/rtl/csrng_block_encrypt.sv:108]: FYI @mwbranstad , you might want to add alerts to CSRNG as well. I've noted that you already use sparse state FSM encodings but the FSMs usually go back into an idle state and the alerts are not connected yet.

File: hw/ip/aes/rtl/aes_control.sv
Comment (vogelpi) [inline hw/ip/aes/rtl/aes_control.sv:537]: @msfschaffner would you recommend to add an additional invalid state and explicitly assign the alert in there (and let the default go into that state as well)?

**Comment by imphil** [comment]
Just a point from the side lines: OTBN called these two alerts "fatal" and "recoverable". I know we don't have system-wide guidance yet, but maybe we can organically evolve a shared naming (and understanding) of these alerts. (https://docs.opentitan.org/hw/ip/otbn/doc/index.html#hardware-interfaces). Maybe we can even expand this to the CSR containing the alert cause (FATAL_ALERT_CAUSE in OTBN, a bitfield)?

**Comment by vogelpi** [comment]
Yes, I've noted that, too. In Ibex (not yet connected to alert handler though) we are using `minor` and `major` which is a bit shorter, which is why I went with that. I think it would be great to harmonize this across all IP.

The CSR bitfield in AES is named differently because at some point there will likely come major alerts that are not fatal.

File: hw/ip/csrng/rtl/csrng_block_encrypt.sv
Comment (mwbranstad) [reply hw/ip/csrng/rtl/csrng_block_encrypt.sv:108]: @vogelpi based on this PR intent, it does seem to make sense to capture illegal states and direct them to an alert. Since there are discussions on a consistent naming and implementation scheme, I will follow whatever is defined. Additionally, all of the entropy complex blocks need a general security review for D2 and I would expect topics such as this to come up in that review.

**Comment by imphil** [comment]
On naming: yes, minor and major are shorter, but also have less meaning. This gets "interesting" when also considering that "there will likely come major alerts that are not fatal.". 

In OTBN, we did try to group alerts by the action the system and the software developer is expected to take: on a fatal alert, a system reset is needed. On a recoverable alert, restarting the operation is enough. 

It seems to me that in the AES case you (will) have three types of alerts: fatal alerts (system reset required), recoverable alerts (AES restart required), and informative alerts (just FYI, operation is continuing as usual). Currently, "fatal" and "recoverable" are both named "major" alerts for AES, with what I called "informative" being a "minor" alert.

@msfschaffner what's your thinking on that?

File: hw/ip/aes/rtl/aes_control.sv
Comment (msfschaffner) [reply hw/ip/aes/rtl/aes_control.sv:537]: Yes, exactly. 
This is also what I have done in LC and OTP.
Glitched FSMs should be fatal, go into a terminal error state and constantly drive out an alert.

See also here for an example:
https://github.com/lowRISC/opentitan/blob/33c9078a70d69971fb6e158d9933d7d74aa70286/hw/ip/otp_ctrl/rtl/otp_ctrl_dai.sv#L594-L619

File: hw/ip/aes/rtl/aes_cipher_control.sv
Comment (msfschaffner) [inline hw/ip/aes/rtl/aes_cipher_control.sv:85]: @cdgori is 3 enough in this case or should we move to 4/5?

File: hw/ip/aes/rtl/aes_core.sv
Comment (msfschaffner) [inline hw/ip/aes/rtl/aes_core.sv:608]: should we make minor errors observable via status registers as well?

File: hw/ip/aes/rtl/aes_core.sv
Comment (msfschaffner) [inline hw/ip/aes/rtl/aes_core.sv:606]: This is related to your previous question: Fatal conditions such as register storage or FSM state corruption should cause the alert to be constantly asserted. For FSMs this can be achieved by adding a terminal error state as discussed in the other comment. For the storage error, you may have to introduce another flop to latch the error.

**Review by msfschaffner — COMMENTED** [review]
Hey @imphil: your summary on alert grouping is correct. I actually meant to bring this up in our next silicon meeting. We have converged on the alert usage in this issue https://github.com/lowRISC/opentitan/issues/3767, but haven't had a chance to formalize this and agree on the naming scheme. As of now, I am also leaning towards the more descriptive naming (fatal, recoverable), but we can discuss that.

-------------------------------

RE your changes, @vogelpi: they LGTM, modulo the comments.

**Comment by silvestrst** [comment]
@silvestrst for visibility

File: hw/ip/aes/rtl/aes_control.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_control.sv:537]: Yes I noted that. But in OTP, the FSM can also go into this error state through regular state transitions, e.g. by evaluating some input signal combinations. In AES, the error state can only be reached through bit flips, i.e. if the state signal takes an invalid value in which case we end up in the default state. So, my impression is that having in the default would probably be fine as well. Or is there another difference?

Anyway, I will probably add an explicit error state as suggested.

Comment (vogelpi) [reply hw/ip/aes/rtl/aes_control.sv:537]: Actually, there is another difference: If the device just ends up in the default state. The adversary could try to inject another bit flip, either moving the device back into the original state or just a little bit closer to the target state. So, I definitely have to add a real error state.

File: hw/ip/aes/rtl/aes_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv:606]: I will add an error state as discussed in the other conversation. As for the storage error: Once a storage error is detected, the corresponding register cannot be written anymore by SW. However, this doesn't prevent the injection of further bit flips until the registers match again. So yes, I have to add another flop. Thanks for pointing that out!

File: hw/ip/aes/rtl/aes_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv:608]: So far, my view was that we don't need or even want that in some cases:
- Minor errors like shadow reg update errors go away automatically. SW anyway needs to read back the shadow reg to see if the write was successful. 
- For some errors (e.g. induced through fault injection attacks), it seems better not to let software know as this information alone can already be very useful for breaking the device.

However, a couple of people have been asking for this now. It's probably better to add the bit and then exclude errors from showing up in there if it comes to defending against some fault attacks.

**Comment by vogelpi** [comment]
I like the classification and naming scheme of @imphil and will go with that :+1:

File: hw/ip/aes/rtl/aes_core.sv
Comment (msfschaffner) [reply hw/ip/aes/rtl/aes_core.sv:608]: Yes I see your point. 
This may be something we want to discuss and audit on the system level later on. 
There are several other modules (including OTP) that currently output more detailed alert cause info.
Maybe it makes sense to guard these cause bits by a life cycle signal at some point in order to only make them accessible in DEV, RMA and TEST_UNLOCKED, for instance.
@cdgori @tjaychen

**Comment by msfschaffner** [comment]
Another thing I wanted to mention: If your state machine is going to have an explicit and terminal error state, you may also want to consider moving it into that state upon assertion of `lc_escalate_en`, which is the life cycle signal that triggers state wiping features.

File: hw/ip/aes/rtl/aes_control.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_control.sv]: This is done now. What is missing is going into the terminal error state upon receiving the life cycle escalation signal. This might also need clearing of registers and is out of scope for this PR. But I added this item to the AES tracking issue.

File: hw/ip/aes/rtl/aes_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv]: I have added a second bit in the status register to inform about update errors (recoverable). Adding a cause register to give more details can be done but it should first be discussed if we really want that. I have added this item to my AES tracking issue.

File: hw/ip/aes/rtl/aes_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv]: This is done now as well: Storage errors remain asserted until reset. Update errors remain asserted until the next write to the control register (restarts operation).

**Comment by vogelpi** [comment]
Update: I've switched the alert naming system as discussed in yesterday's meeting and addressed also all other comments. Some things are IMO out of scope for this PR but I've added them to my AES tracking issue.

File: hw/ip/aes/data/aes.hjson
Comment (msfschaffner) [inline hw/ip/aes/data/aes.hjson:151]: I was actually going to propose a naming scheme that uses prefixes, e.g.:
`fatal_ctrl_err_storage`, `recov_ctrl_err_update`

That way it is still possible to further differentiate the alerts - but we know that they will always start with either of the two prefixes.

WDYT?

Comment (cdgori) [reply hw/ip/aes/data/aes.hjson:151]: I think the prefix is a good idea.

File: hw/ip/aes/rtl/aes_cipher_control.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_cipher_control.sv:365]: I am generally in favor of these terminal error states (and the default -> ERROR transition) since they generally imply something (quite) bad has happened.

However, I will note for posterity that they do increase the overall "brittleness" of the design - and also that we will reach these terminal error states with naturally occurring alpha-particle flips - i.e. we give up some amount of fault tolerance / fault resiliency in order to catch malicious attacks.  I see no way around this problem if we want to consider glitch attacks as severe enough to force a reset, but figure I should just mention it now as a potential downside to the approach.

File: hw/ip/aes/rtl/aes_control.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_control.sv:126]: I think anything HD>2 is fine here.

**Review by cdgori — APPROVED** [review]
LGTM.  

I'll approve now, if we decide to change the naming scheme for the alert signals, you can do it in this PR or a follow-up.

File: hw/ip/aes/rtl/aes_core.sv
Comment (cdgori) [reply hw/ip/aes/rtl/aes_core.sv]: If we supply cause info (here or elsewhere), I'm in favor of guarding it with LC (DEV | RMA | TEST_UNLOCKED sounds right to me).

We should decide more generally if we'll supply cause info and be consistent.

File: hw/ip/aes/data/aes.hjson
Comment (vogelpi) [reply hw/ip/aes/data/aes.hjson:151]: I like the idea of the prefix but in this particular case I didn't use it for the following reason.

Before this PR, the two alerts were `ctrl_err_update` and `ctrl_err_storage`. There was just one source (the shadow register) that could create alerts. Now with this PR we have multiple sources: shadow register, FSM states, and more will follow (handshake signals, mux selector signals, round counter etc.). Thus I don't think it makes sense to mention any more the alert source in the name but instead just call this the `fatal` alert. Then there is the `recoverable` alert which at the moment can only be triggered by the shadow register as before. I don't consider it useful to have a `fatal` alert and a `recov_ctrl_err_update`. So it's just the `recoverable` alert. Also here we might add more sources soon.

For the guidance the document, my suggestion would be to say something like: "It is recommended to prefix alert names with either `fatal` or `recov/recoverable` to indicate the severeness of the alert and possible implications inside the peripheral (this would be the blue feedback loop in your slide). If a peripheral collects multiple alert sources in just 1 alert per group (`fatal`, `recoverable`) the recommendation is to only use the prefix." In the latter case we can then recommend to add a cause register for providing more information on the actual alert cause, guarded by LC as discussed in another comment.

Side note: For most peripherals I think we want to reduce the number of alerts to a minimum of 2 or 3 to avoid problems like we see for interrupts (see #3421). So maybe the guidance should be: try to expose just one fatal and one recoverable alert. If more alerts are needed, prefix your alerts with either `fatal` or `recov`.

Do you think this is reasonable @msfschaffner ?

File: hw/ip/aes/rtl/aes_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv]: Yes, I agree. I've restarted the discussion in #3767 but we can also fan out to a new issue if preferred.

**Comment by vogelpi** [comment]
Thanks @cdgori for reviewing and approving this PR.

In this PR we've triggered some good discussions but I don't think we can solve everything in here. More thought is needed to work out some more general guidance on alerts. I am fine to rework the alerts inside AES once we have this guidance but would like to move forward with this PR (and others that depend on it). I'll leave the final word to @msfschaffner .

File: hw/ip/aes/data/aes.hjson
Comment (msfschaffner) [reply hw/ip/aes/data/aes.hjson:151]: Your guidance suggestion sounds good to me @vogelpi, I'll try to incorporate this into the documentation (PR will come soon).

Ok to move forward now and amend later (if needed)!

**Review by msfschaffner — APPROVED** [review]

File: hw/ip/aes/data/aes.hjson
Comment (vogelpi) [reply hw/ip/aes/data/aes.hjson:151]: Thanks @msfschaffner for the thorough review! I am merging this now.

**Pull request merged**
